### PR TITLE
Provide a global way to customize QPS server builder options

### PR DESCRIPTION
**Pull request by apolcyn**
Needed to test servers internally that need to use custom `ServerBuilderOptions` on all servers created in a QPS benchmark.

Provides a configurable `g_create_qps_server_builder` function that defaults to `new ServerBuilder()`  but can be overridden.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,427      Total (>)     10,770,426

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,423      Total (<)     10,770,425

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,428      Total (>)     10,770,418

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,416      Total (<)     10,770,419

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,423      Total (<)     10,770,426

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,423      Total (>)     10,770,422

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: test/cpp/qps/qps_server_builder.cc [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+#include "qps_server_builder.h"
+
+using grpc::ServerBuilder;
+
+namespace grpc {
+namespace testing {
+
+std::unique_ptr<ServerBuilder> DefaulCreatetQpsServerBuilder() {
```
Comment (ncteisen) [inline test/cpp/qps/qps_server_builder.cc]: Put in anonymous namespace

File: test/cpp/qps/qps_server_builder.cc [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  return std::unique_ptr<ServerBuilder>(new ServerBuilder());
+}
+
+std::function<std::unique_ptr<ServerBuilder>()> g_create_qps_server_builder =
```
Comment (ncteisen) [inline test/cpp/qps/qps_server_builder.cc]: Put in anonymous namespace

File: test/cpp/qps/qps_server_builder.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+#include <memory>
+
+#include <grpcpp/server_builder.h>
+
+namespace grpc {
+namespace testing {
+
+// CreateQpsServerBuilder creates a new ServerBuilder, using
+// the "create ServerBuilder" func that was set in SetCreateQpsServerBuilderFunc
+// if one has been set.
```
Comment (ncteisen) [inline test/cpp/qps/qps_server_builder.h]: Add note about default, unset behavior

File: test/cpp/qps/qps_server_builder.cc (hunk 1 above)
Comment (apolcyn) [reply test/cpp/qps/qps_server_builder.cc]: done

File: test/cpp/qps/qps_server_builder.cc (hunk 2 above)
Comment (apolcyn) [reply test/cpp/qps/qps_server_builder.cc]: done

File: test/cpp/qps/qps_server_builder.h (hunk 3 above)
Comment (apolcyn) [reply test/cpp/qps/qps_server_builder.h]: done

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,976,225      Total (=)      1,976,225

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,770,424      Total (>)     10,770,423

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by apolcyn** [comment]
Test failures:TSAN C++ - https://github.com/grpc/grpc/issues/16103

interop cloud to cloud: https://github.com/grpc/grpc/issues/16078

**Pull request merged**